### ISSUE TITLE
Follow-up work for `download-rustc`

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#81930]
These are all the features for `download-rustc` that didn't land in the initial MVP (https://github.com/rust-lang/rust/pull/79540).

## Required before this is made a default

- [x] Download a separate beta compiler for stage0 and only use CI artifacts for stage1 onward. Right now, every time you rebase, stage0 tools (bootstrap, tidy, ...) will have to be recompiled. Additionally running `x.py setup tools` will compile rustbuild twice. https://github.com/rust-lang/rust/pull/79540#issuecomment-757047321, https://github.com/rust-lang/rust/pull/82739
- [x] Compile a new version of tracing so that rustdoc still gets debug logging (since CI artifacts always disable `debug` and `trace` logging). https://github.com/rust-lang/rust/pull/81932
- [x] Right now only rustdoc is ever rebuilt. This is not ideal and should probably at least compile compiler tools (rustfmt, clippy, miri). https://github.com/rust-lang/rust/pull/79540#issuecomment-775634693, https://github.com/rust-lang/rust/pull/82480
- [x] Using `git log --author=bors` sometimes breaks. This should use `git merge-base` instead, or use some other workaround that doesn't break. https://github.com/rust-lang/rust/pull/79540#discussion_r572572280, https://github.com/rust-lang/rust/pull/82740
- [x] This breaks during release week (but only release week). #83350

## Nice to have

- [x] Add `download-rustc = "if-unchanged"` which rebuilds rustc instead of warning when `compiler/` has been modified. https://github.com/rust-lang/rust/issues/81930#issuecomment-785527408, https://github.com/rust-lang/rust/pull/83368
- [x] `x.py check --stage 0` doesn't work. https://github.com/rust-lang/rust/issues/81930#issuecomment-805428682, https://github.com/rust-lang/rust/pull/82739
- [x] `x.py check --stage 1 src/tools/rustdoc` doesn't work. https://github.com/rust-lang/rust/issues/81930#issuecomment-813408485, fixed in https://github.com/rust-lang/rust/pull/110121
- [x] Add `x.py setup tools` to enable this conveniently (it doesn't make sense to use this for compiler developers). This shouldn't be added until people have used this feature for a little bit and it works reasonably well. https://github.com/rust-lang/rust/pull/83370
- [ ] Enable verification for the commit tarballs. This requires first signing nightly commits; right now the .sha256 URLs gave a 404. This seems fairly simple to fix but also not high priority.
- [x] Give a better error message when rebasing an old branch (I think commits are kept at most 160 days?). https://github.com/rust-lang/rust/pull/79540#issuecomment-751481291
- [x] It would be nice to support cross-compiling the standard library - fixed in https://github.com/rust-lang/rust/pull/110113
- [x] Some tests fail when download-rustc is enabled: https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AA-testsuite+label%3AA-download-rustc
- ~~Combine `download-rustc` and `download-ci-llvm` into the same option. This needs quite a bit more design work. https://github.com/rust-lang/rust/pull/79540#issuecomment-775477288, https://github.com/rust-lang/rust/issues/83369~~ This doesn't seem useful.

**Comment by jyn514** [comment]
> Right now only rustdoc is ever rebuilt. This is not ideal and should probably at least compile compiler tools (rustfmt, clippy, miri). #79540 (comment)

While working on this, I found that the handling of `ENABLE_DOWNLOAD_STAGE0` is broken: if a `Step` is called as part of `ensure()` instead of through `maybe_run`, it will still be built even if the constant is set to false. Not yet sure how to fix this.

<details>

Note that this has local changes enabling the constant for clippy:

```diff
diff --git a/src/bootstrap/tool.rs b/src/bootstrap/tool.rs
index 5c874f69bd9..c82b78c1c71 100644
--- a/src/bootstrap/tool.rs
+++ b/src/bootstrap/tool.rs
@@ -564,6 +564,7 @@ pub struct Cargo {
 impl Step for Cargo {
     type Output = PathBuf;
     const DEFAULT: bool = true;
+    const ENABLE_DOWNLOAD_RUSTC: bool = true;
     const ONLY_HOSTS: bool = true;
 
     fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
@@ -647,6 +648,7 @@ pub struct $name {
         impl Step for $name {
             type Output = Option<PathBuf>;
             const DEFAULT: bool = true; // Overwritten below
+            const ENABLE_DOWNLOAD_RUSTC: bool = true;
             const ONLY_HOSTS: bool = true;
 
             fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
```

```
$ x.py build -v src/tools/clippy
detected default triple x86_64-unknown-linux-gnu
Updating only changed submodules
Submodules updated in 0.01 seconds
using downloaded stage1 artifacts from CI (commit fe1bf8e05c39bdcc73fc09e246b7209444e389bc)
running: /home/joshua/rustc3/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /home/joshua/rustc3/src/bootstrap/Cargo.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
running: /home/joshua/rustc3/build/bootstrap/debug/bootstrap build -v src/tools/clippy
finding compilers
CC_x86_64-unknown-linux-gnu = "cc"
CFLAGS_x86_64-unknown-linux-gnu = ["-ffunction-sections", "-fdata-sections", "-fPIC", "-m64"]
CXX_x86_64-unknown-linux-gnu = "c++"
CXXFLAGS_x86_64-unknown-linux-gnu = ["-ffunction-sections", "-fdata-sections", "-fPIC", "-m64"]
AR_x86_64-unknown-linux-gnu = "ar"
running sanity check
auto-detected local-rebuild 1.52.0-nightly
learning about cargo
> Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  > Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  > Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
< Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
> Clippy { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, extra_features: [] }
  > ToolBuild { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "clippy-driver", path: "src/tools/clippy", mode: ToolRustc, is_optional_tool: true, source_type: InTree, extra_features: [] }
    > Rustc { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      > Std { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        > StartupObjects { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        < StartupObjects { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        > Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
          c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        < Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        > StdLink { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
          c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
          c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        < StdLink { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      < Std { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Std { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      > Llvm { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      < Llvm { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      > RustcLink { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      < RustcLink { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
    < Rustc { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < ToolBuild { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "clippy-driver", path: "src/tools/clippy", mode: ToolRustc, is_optional_tool: true, source_type: InTree, extra_features: [] }
< Clippy { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, extra_features: [] }
c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
> CargoClippy { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, extra_features: [] }
  > ToolBuild { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "cargo-clippy", path: "src/tools/clippy", mode: ToolRustc, is_optional_tool: true, source_type: InTree, extra_features: [] }
    c Rustc { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
    c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < ToolBuild { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "cargo-clippy", path: "src/tools/clippy", mode: ToolRustc, is_optional_tool: true, source_type: InTree, extra_features: [] }
< CargoClippy { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, extra_features: [] }
> Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  > Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  > Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
  < Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
< Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
> Clippy { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, extra_features: [] }
  > ToolBuild { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "clippy-driver", path: "src/tools/clippy", mode: ToolRustc, is_optional_tool: true, source_type: InTree, extra_features: [] }
    > Rustc { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
      > Std { target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        > StartupObjects { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        < StartupObjects { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Assemble { target_compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        > Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
          c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        < Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Libdir { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
        c Sysroot { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/joshua/rustc3/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "8" "--release" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/home/joshua/rustc3/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
   Compiling core v0.0.0 (/home/joshua/rustc3/library/core)
^C  Building [======>                     ] 9/33: core                              
Build completed unsuccessfully in 0:00:01
```

</details>

**Comment by jyn514** [comment]
lol I also found that this completely ignores all test suites, making it pretty useless:
```
Not running bootstrap::test::RustdocUi because its artifacts have been downloaded from CI (`download-rustc` is set)
```

**Comment by camelid** [comment]
> Required before this is made a default

I don't think this should be enabled by default until [this](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/.60download-rustc.60.20is.20open.20for.20business!/near/227600115) is addressed as well.

**Comment by jyn514** [comment]
@camelid that's why it has a warning. Do you think it should force you to disable the option? That seems like it could be frustrating.

**Comment by jyn514** [comment]
I definitely don't think it should silently use the in-tree compiler if you make changes, it should always do what you specify in config.toml.

**Comment by camelid** [comment]
I was thinking that there could be a `download-rustc = "if-no-local-changes"` option or something like that, where it would only download rustc if `compiler/` isn't dirty.

**Comment by jyn514** [comment]
Ah ok, that seems useful. I don't think it needs to block turning this on by default, though.

**Comment by camelid** [comment]
> Ah ok, that seems useful. I don't think it needs to block turning this on by default, though.

Won't this be annoying to new rustc contributors who have to dig around in their settings?

**Comment by jyn514** [comment]
@camelid I found the confusion. When I say "by-default", I mean that I want to add a new `x.py setup tools` option that's specifically intended for rustdoc/clippy/rustfmt contributors and has `download-rustc = true`. I don't want to turn it on for the whole repository, that wouldn't make sense (or at least, not until `if-no-changes` is implemented).

**Comment by camelid** [comment]
Ah, I had a feeling we were talking past each other :)

Sounds good to me, as long as the warning is helpful enough that if people start doing compiler contributions they don't get stuck.

**Comment by jyn514** [comment]
>  This breaks during release week (but only release week, for some reason). Ran into this again recently so it's not related to the original (broken) implementation. = note: /usr/bin/ld: cannot find -lLLVM-11-rust-1.52.0-nightly

I managed to fix this locally by hard-coding `llvm-sha` to `ed4005da44b0dcbeaa35951ef994da82e156258f` (which updated to 1.53: https://github.com/rust-lang/rust/commit/ed4005da44b0dcbeaa35951ef994da82e156258f). @Mark-Simulacrum how hard would it be to add 'bump `src/bootstrap/download-ci-llvm-stamp`' to the release process?

**Comment by jyn514** [comment]
Actually, maybe a better fix is to find which commit most recently modified `src/version`, and use that if it's more recent than `llvm_sha`.

**Comment by camelid** [comment]
FYI, `x.py check rustdoc` doesn't seem to work properly. It starts checking stage0 std artifacts and then stage0 compiler artifacts, and then errors out with this message:

```
Checking stage0 compiler artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling autocfg v1.0.0
   Compiling proc-macro2 v1.0.19
    Checking cfg-if v0.1.10
   Compiling libc v0.2.88
error[E0464]: multiple matching crates for `core`
  |
  = note: candidates:
          crate `core`: /Users/user/rust2/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libcore-72787c1a9e4c6b7c.rlib

error: aborting due to previous error

error: could not compile `cfg-if`
```

Weirdly, it only lists one `core` candidate yet it says there are multiple matching crates.

`x.py build` works fine.

I don't know if this is a glitch on my system (hopefully `x.py clean` won't be required) or a bug in bootstrap, but thought I'd report it in case it's a bug.

**Comment by jyn514** [comment]
Huh, I just realized that `build --stage 0` works currently, I wouldn't expect it to (it will break when https://github.com/rust-lang/rust/pull/82739 lands which is why I made --stage 2 the default in that PR when `download-rustc` is set). I think changing the default to `check --stage 2` should work well enough.

Alternatively I can try to mess with not copying libstd artifacts into stage0-sysroot when download-rustc is set, but that sounds more complicated.

**Comment by jyn514** [comment]
> I think changing the default to check --stage 2 should work well enough.

Oh, so close.
```
error[E0464]: multiple matching crates for `lazy_static`
  --> src/librustdoc/lib.rs:22:1
   |
22 | extern crate lazy_static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidates:
           crate `lazy_static`: /home/joshua/rustc3/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblazy_static-7d4fcba0e28517d0.rlib
```

**Comment by jyn514** [comment]
Huh, #82739 seems to have also fixed `x.py check --stage 0`.

**Comment by jyn514** [comment]
`x.py check --stage 1 src/tools/rustdoc` seems to still be broken. But that seems weird to need, almost nothing in rustdoc should need cfg(bootstrap).

**Comment by jyn514** [comment]
>  It would be nice to support cross-compiling the standard library. Right now this gives an assertion failure I think.

Err it looks like `x.py build library/std` always does nothing. So --target is also ignored.

**Comment by jyn514** [comment]
Oh huh, this is an unexpected side-effect: since https://github.com/rust-lang/rust/pull/82739, building compiletest also requires building stage0 std first. I think using master libstd intentional since https://github.com/rust-lang/rust/issues/59264? I don't think it necessarily needs to change, maybe just documented somewhere?

**Comment by camelid** [comment]
Reposting from [this Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/confusing.20download-rustc.20behavior) as requested by `@jyn514`:

I just ran into something confusing with `download-rustc`: normal `x.py build` will build rustdoc and put it in **stage1**. However, `x.py build` with `download-rustc` enabled will put it in **stage2**. This means that `rustdoc +stage1 ...` won't run my local, modified version of rustdoc if `download-rustc` is set. Specifically, `rustdoc +stage1 ...` seems to run *nightly* rustdoc (or something static).

I'm guessing this is because rustdoc built with `download-rustc` is being built with stage2 rustc, thus it is stage2 rustdoc, whereas rustdoc built locally *without* `download-rustc` is usually built with stage1 rustc, thus it is stage1 rustdoc. However, this behavior was very confusing to me and it took me a bunch of experimenting to figure out why my changes weren't showing up in the rustdoc output.

I imagine it could be even more confusing for new contributors since we recommend using stage1 builds, but then when they use `download-rustc` they have to remember that it's actually a *stage2* build and adjust their commands accordingly. Plus, it's frustrating even if you know what's going on to switch between stage1 and stage2 depending on whether you're working on rustc or rustdoc (or a different tool).

Is there any way to solve this problem? I imagine there are multiple possible solutions, some better than others.

**Comment by jyn514** [comment]
> I just ran into something confusing with `download-rustc`: normal `x.py build` will build rustdoc and put it in **stage1**. However, `x.py build` with `download-rustc` enabled will put it in **stage2**. This means that `rustdoc +stage1 ...` won't run my local, modified version of rustdoc if `download-rustc` is set. Specifically, `rustdoc +stage1 ...` seems to run _nightly_ rustdoc (or something static).

@bjorn3 suggests deleting the `stage1` and `stage0-sysroot` directories when using download-rustc, to avoid caching issues like this. `rustdoc +stage1` still won't do the right thing, but it will at least give a hard error instead of silently using the wrong version.

**Comment by camelid** [comment]
Great idea! That would be very helpful.

**Comment by jyn514** [comment]
> > I just ran into something confusing with `download-rustc`: normal `x.py build` will build rustdoc and put it in **stage1**. However, `x.py build` with `download-rustc` enabled will put it in **stage2**. This means that `rustdoc +stage1 ...` won't run my local, modified version of rustdoc if `download-rustc` is set. Specifically, `rustdoc +stage1 ...` seems to run _nightly_ rustdoc (or something static).
> 
> bjorn3 suggests deleting the `stage1` and `stage0-sysroot` directories when using download-rustc, to avoid caching issues like this. `rustdoc +stage1` still won't do the right thing, but it will at least give a hard error instead of silently using the wrong version.

Done in https://github.com/rust-lang/rust/commit/973ff033f38a1ee23798560e748a7e2035f49ed0

**Comment by jyn514** [comment]
> Give a better error message when rebasing an old branch (I think commits are kept at most 160 days?). https://github.com/rust-lang/rust/pull/79540#issuecomment-751481291

I'm going to cross this off the list - right now download-ci-llvm gives a useful error first and I think it's very rare to want to download rustc but not LLVM.

```
downloading https://ci-artifacts.rust-lang.org/rustc-builds/8ce3204af9463db3192ea1eb31c45c2f6d4b5ae6/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404                                                                                

error: failed to download llvm from ci

help: old builds get deleted after a certain time
help: if trying to compile an old commit of rustc, disable `download-ci-llvm` in config.toml:

[llvm]
download-ci-llvm = false
```

**Comment by jyn514** [comment]
> Some tests fail when download-rustc is enabled: https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AA-testsuite+label%3AA-download-rustc

this is now fixed well enough that `x test` by default succeeds :) `x test --stage 1 rustdoc` is still broken, but that should be extremely rare to need to run, and I don't think we need a whole tracking issue for it.
